proof_system/src/sub_protocols/bbs_plus.rs

```diff
@@ -44,33 +41,69 @@ impl<'a, E: Pairing> PoKBBSSigG1SubProtocol<'a, E> {
         &mut self,
         rng: &mut R,
         blindings: BTreeMap<usize, E::ScalarField>,
-        mut witness: crate::witness::PoKBBSSignatureG1<E>,
+        witness: crate::witness::PoKBBSSignatureG1<E>,
     ) -> Result<(), ProofSystemError> {
         if self.protocol.is_some() {
             return Err(ProofSystemError::SubProtocolAlreadyInitialized(self.id));
         }
         // Create messages from revealed messages in statement and unrevealed in witness
-        let mut messages = Vec::with_capacity(self.signature_params.supported_message_count());
-        let mut revealed_indices = BTreeSet::new();
-        for i in 0..self.signature_params.supported_message_count() {
-            if witness.unrevealed_messages.contains_key(&i) {
-                messages.push(witness.unrevealed_messages.remove(&i).unwrap());
-            } else if self.revealed_messages.contains_key(&i) {
-                revealed_indices.insert(i);
-                messages.push(*self.revealed_messages.get(&i).unwrap());
-            } else {
-                return Err(ProofSystemError::BBSPlusProtocolMessageAbsent(self.id, i));
-            }
-        }
+        let mut invalid_blinding_idx = None;
+        let messages_to_commit_with_blindings = witness
+            .unrevealed_messages
+            .iter()
+            .merge_join_by(blindings, |(&m_idx, _), (b_idx, _)| m_idx.cmp(b_idx))
+            .scan((), |(), either| {
+                let item = match either {
+                    EitherOrBoth::Left((idx, msg)) => {
+                        (*idx, MessageOrBlinding::BlindMessageRandomly(msg))
+                    }
+                    EitherOrBoth::Both((idx, message), (_, blinding)) => (
+                        *idx,
+                        MessageOrBlinding::BlindMessageWithConcreteBlinding { message, blinding },
+                    ),
+                    EitherOrBoth::Right((idx, _)) => {
+                        invalid_blinding_idx.replace(idx);
+
+                        return None;
+                    }
+                };
+
+                Some(item)
+            });
+
+        let revealed_messages = self
+            .revealed_messages
+            .iter()
+            .map(|(idx, msg)| (*idx, MessageOrBlinding::RevealMessage(msg)));
+
+        let mut invalid_message_idx = None;
+        let all_messages = messages_to_commit_with_blindings
+            .merge_by(revealed_messages, |(a, _), (b, _)| a <= b)
+            .scan(None, |last, (idx, msg)| {
+                if last.map_or_else(|| idx != 0, |last| last + 1 != idx) {
```

Comment:
Can you please explain what this is for? That there are no gaps in indices returned by `merge_by`?

Shouldn't chaining the iterator `messages_to_commit_with_blindings` and `revealed_messages` be sufficient? The check that all messages indices are covered could be made by comparing the size of the result with  `self.revealed_messages.len() + self.unrevealed_messages.len()`

Reply:
Same inconsistent indices problem as described above: we have `revealed_messages` and `unrevealed_messages` which are both mappings from the message index to the message, and either can have non-sequential indices.

Reply:
The check here makes sense. Thanks.